### PR TITLE
Adding DS3+LoRA support to Falcon and Llama models for Text-generation

### PR DESCRIPTION
**Pull request by skanakamedal**

### CODE REVIEW

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py [hunk 1]
```diff
@@ -98,7 +99,10 @@
 ]
 
 
-FORCE_4BIT_QUANTIZATION_MODELS = ["Llama-2-70b", "tiiuae-falcon-40b"]
+FORCE_4BIT_QUANTIZATION_MODELS = [
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py:102]: I believe, we can remove this check?
cc @amangupta26

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py [hunk 2]
```diff
@@ -504,6 +579,30 @@ def identify_deepspeed_stage(deepspeed_config_json: Dict[str, Any]) -> int:
     return ds_stage
 
 
+def validate_setting_value(source: Namespace, key: str, value: Union[List, str]) -> bool:
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py:582]: Consider changing `validate_setting_value --> is_match` and change signature accordingly

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py [hunk 3]
```diff
@@ -532,8 +635,10 @@ def check_for_invalid_ds_zero3_settings(args: Namespace):
         invalid_settings = setting["invalid_settings"]
         fail_run = setting["fail_run"]
         valid_settings = setting["valid_settings"]
-        if all([getattr(args, key, None) == value for key, value in invalid_settings.items()]):
+        if all([validate_setting_value(args, key, value) for key, value in invalid_settings.items()]):
+            logger.info(f"All are True - {fail_run}")
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py:639]: remove logs

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py [hunk 4]
```diff
@@ -652,11 +730,25 @@ def finetune(args: Namespace):
...
 
-    # Enable QLoRA finetune for Llama-70B
+    # init deepspeed config params
+    ds_config_json = None
+    ds_stage = None
+    # Deepspeed enabled
+    if args.apply_deepspeed:
+        # set deepspeed file to be used
+        setup_deepspeed(args)
+        if args.deepspeed:
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py:740]: may be the check is not needed as we are already checking it in L737

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py [hunk 5]
```diff
@@ -817,6 +909,13 @@ def finetune(args: Namespace):
     args.save_strategy = args.evaluation_strategy
     args.save_steps = args.eval_steps
 
+    if args.task_name in [Tasks.NLP_MULTICLASS, Tasks.NLP_MULTILABEL, Tasks.NLP_NER]:
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py:912]: Change this to a function _def setup_automl_nlp(args) -> None_

File: assets/training/finetune_acft_hf_nlp/src/model_selector/model_selector.py [hunk 6]
```diff
@@ -161,7 +161,7 @@ class ModelImportConstants:
                     "trust_remote_code": True,
                 },
                 "tokenizer_config": {
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/model_selector/model_selector.py:163]: remove tokenizer_config

**Review by jpmann — CHANGES_REQUESTED** [review]

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py (hunk 1 above)
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py]: We are not forcing Quantization anymore from script

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py [hunk 7]
```diff
@@ -119,8 +123,28 @@
...
+DEEPSPEED_STAGE3_SUPPORTED_MODEL_TYPES = [
+    HfModelTypes.LLAMA,
+    HfModelTypes.FALCON,
+]
+DEEPSPEED_STAGE3_SUPPORTED_MODEL_TYPES_REGEX_LIST = "|".join(DEEPSPEED_STAGE3_SUPPORTED_MODEL_TYPES)
+# the below regex exludes DEEPSPEED_STAGE3_SUPPORTED_MODEL_TYPES and matches other words
+DEEPSPEED_STAGE3_SUPPORTED_MODEL_TYPES_REGEX = f"^(?!({DEEPSPEED_STAGE3_SUPPORTED_MODEL_TYPES_REGEX_LIST})$)(\\w*)"
+
+
 GRADIENT_CHECKPOINTING_SUPPORTED_MODEL_TYPES = [
```
[the remark this answers is not in the document]
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py]: changed

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py [hunk 8]
```diff
@@ -504,6 +579,30 @@ def identify_deepspeed_stage(deepspeed_config_json: Dict[str, Any]) -> int:
...
+        if re_match is not None:
+            # if there is a regex match then value is matched with value in source
+            is_valid_value = True
+            logger.info(f"Regex matched for {key}.")
+        else:
+            # if there is no regex match
+            is_valid_value = False
+            logger.info(f"Regex did not match for {key}.")
+    else:
+        is_valid_value = source_value == value
```
[the remark this answers is not in the document]
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py]: changed

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py]: changed

Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py]: changed

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py (hunk 3 above)
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py]: removed

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py [hunk 9]
```diff
@@ -119,8 +123,28 @@
...
+
+DEEPSPEED_STAGE3_SUPPORTED_TASKS = [
+    Tasks.TEXT_GENERATION,
+]
+DEEPSPEED_STAGE3_SUPPORTED_TASKS_REGEX_LIST = "|".join(DEEPSPEED_STAGE3_SUPPORTED_TASKS)
+# the below regex exludes DEEPSPEED_STAGE3_SUPPORTED_TASKS and matches other words
+DEEPSPEED_STAGE3_SUPPORTED_TASKS_REGEX = f"^(?!({DEEPSPEED_STAGE3_SUPPORTED_TASKS_REGEX_LIST})$)(\\w*)"
+
+
+DEEPSPEED_STAGE3_SUPPORTED_MODEL_TYPES = [
```
[the remark this answers is not in the document]
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py:130]: For now we are restricting to these model types as tested E2E

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py (hunk 4 above)
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py]: reverted functionality

File: assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py (hunk 5 above)
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py]: added

File: assets/training/finetune_acft_hf_nlp/src/model_selector/model_selector.py (hunk 6 above)
Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/model_selector/model_selector.py]: removed

File: assets/training/finetune_acft_hf_nlp/src/model_selector/model_selector.py [hunk 10]
```diff
@@ -160,9 +160,6 @@ class ModelImportConstants:
                 "model_hf_load_kwargs": {
                     "trust_remote_code": True,
                 },
-                "tokenizer_config": {
```
Comment (amangupta26) [inline assets/training/finetune_acft_hf_nlp/src/model_selector/model_selector.py:163]: Why we removed this?

Comment (skanakamedal) [reply assets/training/finetune_acft_hf_nlp/src/model_selector/model_selector.py:163]: It is not required for latest falon models with model_type=falcon, needed for older falcon models where model_type=RefinedWebModel

**Review by jpmann — APPROVED** [review]

**Review by amangupta26 — APPROVED** [review]

**Review by aggarwal-k — APPROVED** [review]

**Pull request merged**